File: content/examples/other/leetcode/problems/sum-of-two-integers/example-1/solution.js
function getSumBitByBit(a, b) {
  a |= 0;
  b |= 0;
  let result = 0;
  let carry = 0;
  for (let i = 0; i < 32; i += 1) {
    const abit = (a >>> i) & 1;
    const bbit = (b >>> i) & 1;
    const sum = abit ^ bbit ^ carry;
    carry = (abit & bbit) | (abit & carry) | (bbit & carry);
    result |= sum << i;
  }
  return result | 0;
}

if (require.main === module) {
  console.log(getSumBitByBit(1, 2));
  console.log(getSumBitByBit(-4, 7));
}

module.exports = { getSumBitByBit };

File: content/examples/other/leetcode/problems/sum-of-two-integers/example-1/Solution.java
public class Solution {
  static int getSumBitByBit(int a, int b) {
    int result = 0;
    int carry = 0;
    for (int i = 0; i < 32; i += 1) {
      int abit = (a >>> i) & 1;
      int bbit = (b >>> i) & 1;
      int sum = abit ^ bbit ^ carry;
      carry = (abit & bbit) | (abit & carry) | (bbit & carry);
      result |= (sum << i);
    }
    return result;
  }

  public static void main(String[] args) {
    System.out.println(getSumBitByBit(1, 2));
    System.out.println(getSumBitByBit(-4, 7));
  }
}
